### PR TITLE
Made the serializer recreate a connection on each save/load attempts

### PR DESCRIPTION
**Pull request by PsyCommando**
* Serializer now open/close connection once per save operation instead of once at server start. Same when loading a save. Might help with weird timeout issues we got going.
* Moved all the SQL query code into the serializer so there's not random SQL everywhere.
* Made constants for SQL table names and serializer variable types.
* Made a set of macro to handle executing and validating the result of a query. So we  can avoid having retyping the same 4 lines a hundred time.
* Made save progress more verbose and calculated the time for more parts of the save process.

## Changelog

:cl:
prefix:
/:cl:

### CODE REVIEW

File: mods/persistence/controllers/subsystems/persistence.dm [hunk 1]
```diff
@@ -21,15 +21,15 @@
...
-	if(query.NextRow())
-		save_exists = text2num(query.item[1]) > 0
-		in_loaded_world = save_exists
+	save_exists = serializer.save_exists()
+	in_loaded_world = save_exists
 
 /datum/controller/subsystem/persistence/proc/SaveWorld()
 	// Collect the z-levels we're saving and get the turfs!
 	to_world_log("Saving [LAZYLEN(SSpersistence.saved_levels)] z-levels. World size max ([world.maxx],[world.maxy])")
+	sleep(5)
```
Comment (NataKilar) [inline mods/persistence/controllers/subsystems/persistence.dm:30]: What's the purpose of these random sleeps throughout this proc?

File: mods/persistence/controllers/subsystems/persistence.dm [hunk 2]
```diff
@@ -46,36 +46,62 @@
...
+		report_progress("Done wiping previous save in [(REALTIMEOFDAY - time_start_wipe) / (1 SECOND)]s.")
+		sleep(5)
 
 		//
 		// 	ACTUAL SAVING SECTION
 		//
+
+		report_progress("Preparing z-levels for save..")
+		sleep(5)
+		var/time_start_zprepare= REALTIMEOFDAY
```
Comment (NataKilar) [inline mods/persistence/controllers/subsystems/persistence.dm]: Nit: missing a space before = here

File: mods/persistence/controllers/subsystems/persistence.dm [hunk 3]
```diff
@@ -145,6 +171,12 @@
 		serializer.Commit()
 		area_holder.areas.Cut()
 
+		report_progress("Z-levels prepared for save in [(REALTIMEOFDAY - time_start_zprepare) / (1 SECOND)]s.")
+		sleep(5)
+
+		report_progress("Saving z-level turfs..")
+		sleep(5)
+		var/time_start_zsave= REALTIMEOFDAY
```
Comment (NataKilar) [inline mods/persistence/controllers/subsystems/persistence.dm]: Ditto, missing a space before =

File: mods/persistence/controllers/subsystems/persistence.dm [hunk 4]
```diff
@@ -179,7 +211,12 @@
 			serializer.CommitRefUpdates()
 
 		index = 1
+		report_progress("Z-levels turfs saved in [(REALTIMEOFDAY - time_start_zsave) / (1 SECOND)]s.")
+		sleep(5)
 
+		report_progress("Saving z-level areas..")
+		sleep(5)
+		var/time_start_zarea= REALTIMEOFDAY
```
Comment (NataKilar) [inline mods/persistence/controllers/subsystems/persistence.dm]: Ditto

File: mods/persistence/modules/world_save/_defines/_serializer.dm [hunk 5]
```diff
@@ -0,0 +1,42 @@
+/*
+	Serialized type names
+*/ 
+#define SERIALIZER_TYPE_NULL		"NULL"
```
Comment (NataKilar) [inline mods/persistence/modules/world_save/_defines/_serializer.dm:4]: The indentation on all the lines that follow are inconsistent.

File: mods/persistence/controllers/subsystems/persistence.dm [hunk 6]
```diff
@@ -460,4 +394,16 @@
...
+	return
+
+/datum/controller/subsystem/persistence/proc/AddToLimbo(var/datum/thing, var/key, var/limbo_type, var/metadata, var/modify = TRUE)
+	return one_off.AddToLimbo(thing, key, limbo_type, metadata, modify)
+
+/datum/controller/subsystem/persistence/proc/RemoveFromLimbo(var/limbo_key, var/limbo_type)
+	return one_off.RemoveFromLimbo(limbo_key, limbo_type)
+
+/datum/controller/subsystem/persistence/proc/DeserializeOneOff(var/limbo_key, var/limbo_type, var/remove_after = TRUE)
+	return one_off.RemoveFromLimbo(limbo_key, limbo_type, remove_after)
```
Comment (NataKilar) [inline mods/persistence/controllers/subsystems/persistence.dm]: DeserializeOneOff here, I'm guessing, not RemoveFromLimbo?

File: mods/persistence/modules/world_save/z_undefine.dm [hunk 7]
```diff
@@ -0,0 +1,21 @@
+#undef SERIALIZER_TYPE_NULL
```
Comment (NataKilar) [inline mods/persistence/modules/world_save/z_undefine.dm]: ~serializer_undefine.dm or something is preferable. The actual define file also probably doesn't need to be in a separate folder.

File: mods/persistence/controllers/subsystems/persistence.dm [hunk 8]
```diff
@@ -209,20 +246,13 @@
 			serializer.Commit() // cleanup leftovers.
 
 		// Insert our z-level remaps.
-		var/list/z_inserts = list()
```
Comment (NataKilar) [inline mods/persistence/controllers/subsystems/persistence.dm:212]: Why was this moved to the sql serializer? the z_inserts aren't really specific to that serializer and it doesn't reference anything related to the serializer itself. I think the serializer.Commit() and CommitRefUpdates() are actually extraneous and can be removed for the former proc, and moved above the z-level remaps section for the latter proc

**Review by NataKilar — CHANGES_REQUESTED** [review]

File: mods/persistence/controllers/subsystems/persistence.dm (hunk 1 above)
Comment (PsyCommando) [reply mods/persistence/controllers/subsystems/persistence.dm:30]: They're not random. They allow the messages to actually get sent before the server freezes while working. Idk if you've seen how all the message are literally pointless because they all show up at once after the save happens currently. I think its some sort of buffer flushing issue with byond.

File: mods/persistence/modules/world_save/z_undefine.dm (hunk 7 above)
Comment (PsyCommando) [reply mods/persistence/modules/world_save/z_undefine.dm]: I put it in a folder because folders come before files in the hierarchy.

File: mods/persistence/controllers/subsystems/persistence.dm (hunk 8 above)
Comment (PsyCommando) [reply mods/persistence/controllers/subsystems/persistence.dm:212]: Because, you put sql code in the subsystem. So if you use another sort of serializer its gonna run sql command when it probably shouldn't. The main goal is to preserve encapsulation. And it also allows to keep defines contained in the serializer code and not exposed everywhere. If its not related to the serializer it shouldn't try to directly interact with the SQL tables that the serializer uses. 

Also, I'm not changing the ordering of things in this pr, or removing existing code. I'm looking to fix the saving issues first and foremost. If there's superfluous stuff, we can handle that later in another PR. I just wanna keep the risks of introducing more problem low. Especially since I'd need to retest everything if I made those changes

File: mods/persistence/controllers/subsystems/persistence.dm (hunk 6 above)
Comment (PsyCommando) [reply mods/persistence/controllers/subsystems/persistence.dm]: That's a good catch!

File: mods/persistence/modules/world_save/z_undefine.dm (hunk 7 above)
Comment (NataKilar) [reply mods/persistence/modules/world_save/z_undefine.dm]: Mhmm, but through my brief glances, I think if you're undefining it in the same folder it's preferable to have matching files like

_world_save.dm
and
~world_save.dm

or something similar.

File: mods/persistence/controllers/subsystems/persistence.dm (hunk 2 above)
Comment (PsyCommando) [reply mods/persistence/controllers/subsystems/persistence.dm]: done

File: mods/persistence/controllers/subsystems/persistence.dm (hunk 3 above)
Comment (PsyCommando) [reply mods/persistence/controllers/subsystems/persistence.dm]: done

File: mods/persistence/controllers/subsystems/persistence.dm (hunk 4 above)
Comment (PsyCommando) [reply mods/persistence/controllers/subsystems/persistence.dm]: done

File: mods/persistence/modules/world_save/_defines/_serializer.dm (hunk 5 above)
Comment (PsyCommando) [reply mods/persistence/modules/world_save/_defines/_serializer.dm:4]: That's some weirdness with github. Not sure why it does that. I need to insert a space after everything so tabs look like they align on github.

File: mods/persistence/modules/world_save/z_undefine.dm (hunk 7 above)
Comment (PsyCommando) [reply mods/persistence/modules/world_save/z_undefine.dm]: Renamed

File: mods/persistence/controllers/subsystems/persistence.dm (hunk 8 above)
Comment (NataKilar) [reply mods/persistence/controllers/subsystems/persistence.dm:212]: Not super certain on this still but the functionality is identical so it's okay for now. Note to self to go back and make the z transposing a bit more robust.

**Review by NataKilar — APPROVED** [review]
